File: rust/tmp-comptes/src/datastore.rs
use std::io::Error;
use std::time::{SystemTime, UNIX_EPOCH};
use chrono::{DateTime, Local, TimeZone};
use postgres::{Client, NoTls};
use crate::models::{Budget, Expense, Income, Tag};

pub struct Datastore {
    pub client: Client,
}

pub fn new_datastore() -> Datastore {
    let client = Client::connect(
        "postgresql://user:pass@localhost:9001/db",
        NoTls).unwrap();
    return Datastore { client };
}

impl Datastore {
    pub fn list_budgets(&mut self) -> Vec<Budget> {
        let mut budgets: Vec<Budget> = Vec::new();
        for row in self.client.query("SELECT * FROM budgets", &[]).unwrap() {
            let date_system: SystemTime = row.get(2);
            let date: DateTime<Local> = system_time_to_date_time(date_system);
            budgets.push(Budget{
                label: row.get(0),
                amount: row.get(1),
                date,
                tag: row.get(3)
            })
        }
        
        budgets
    }

    pub fn add_budget(&mut self, budget: Budget) {
        self.client.query(
            "INSERT INTO budgets (label, amount, date, tag_id) VALUES ($1, $2, $3, $4)",
            &[&budget.label, &budget.amount, &budget.date.to_string(), &budget.tag],
        ).unwrap();
    }

    pub fn list_expenses(&mut self) -> Vec<Expense> {
        let mut expenses: Vec<Expense> = Vec::new();
        for row in self.client.query("SELECT * FROM expenses", &[]).unwrap() {
            let date_system: SystemTime = row.get(2);
            let date: DateTime<Local> = system_time_to_date_time(date_system);
            expenses.push(Expense{
                label: row.get(0),
                amount: row.get(1),
                date,
                budget: row.get(3)
            })
        }

        expenses
    }
    
    pub fn list_incomes(&mut self) -> Vec<Income> {
        let mut incomes: Vec<Income> = Vec::new();
        for row in self.client.query("SELECT * FROM incomes", &[]).unwrap() {
            let date_system: SystemTime = row.get(2);
            let date: DateTime<Local> = system_time_to_date_time(date_system);
            incomes.push(Income{
                label: row.get(0),
                amount: row.get(1),
                date,
            })
        }

        incomes
    }

    pub fn list_tags(&mut self) -> Vec<Tag> {
        let mut tags: Vec<Tag> = Vec::new();
        for row in self.client.query("SELECT * FROM tags", &[]).unwrap() {
            tags.push(Tag{
                label: row.get(0),
                description: row.get(1),
                icon: row.get(2),
            })
        }

        tags
    }
}

fn system_time_to_date_time(t: SystemTime) -> DateTime<Local> {
    // Source : https://users.rust-lang.org/t/convert-std-time-systemtime-to-chrono-datetime-datetime/7684/4
    let (sec, nsec) = match t.duration_since(UNIX_EPOCH) {
        Ok(dur) => (dur.as_secs() as i64, dur.subsec_nanos()),
        Err(e) => { // unlikely but should be handled
            let dur = e.duration();
            let (sec, nsec) = (dur.as_secs() as i64, dur.subsec_nanos());
            if nsec == 0 {
                (-sec, 0)
            } else {
                (-sec - 1, 1_000_000_000 - nsec)
            }
        },
    };
    
    Local.timestamp_opt(sec, nsec).unwrap()
}

File: go/comptes/internal/datastore.go
package internal

import (
	"context"
	"fmt"

	"github.com/gofrs/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Datastore struct {
	dbpool *pgxpool.Pool
}

func NewDatastore(cfg *Config) (*Datastore, error) {
	dbpool, err := pgxpool.New(context.Background(), cfg.Db.Connection)
	if err != nil {
		return nil, fmt.Errorf("could not connect to database: %w", err)
	}
	// defer dbpool.Close()

	return &Datastore{
		dbpool: dbpool,
	}, nil
}

func (d *Datastore) AddBudget(budget *Budget) error {
	query := `INSERT INTO budgets (label, amount, date, tag_id)
VALUES (@label, @amount, @date, @tagId)`
	args := pgx.NamedArgs{
		"label":  budget.Label,
		"amount": budget.Amount,
		"date":   budget.Date,
		"tagId":  budget.TagID.String(),
	}

	_, err := d.dbpool.Exec(context.Background(), query, args)
	if err != nil {
		return fmt.Errorf("error creating budget: %s", err)
		// return err
	}

	return nil
}

func (d *Datastore) ListBudgets(year, month int) (budgets []*Budget, err error) {
	rows, err := d.dbpool.Query(
		context.Background(),
		"select * from budgets where extract(year from date) = $1 and extract(month from date) = $2",
		year,
		month,
	)
	if err != nil {
		return nil, fmt.Errorf("could not list budgets: %w", err)
	}

	for rows.Next() {
		var budget Budget
		err := rows.Scan(
			&budget.ID,
			&budget.Label,
			&budget.Amount,
			&budget.Date,
			&budget.TagID,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan budget: %w", err)
		}

		budgets = append(budgets, &budget)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate budgets: %w", err)
	}

	return
}

func (d *Datastore) GetTagByLabel(tagLabel string) (*Tag, error) {
	var tag Tag
	err := d.dbpool.QueryRow(context.Background(), "select * from tags where label = $1", tagLabel).Scan(&tag)
	if err != nil {
		return nil, fmt.Errorf("could not get tag: %w", err)
	}

	return &tag, nil
}

func (d *Datastore) ListBudgetsForTagId(tagId uuid.UUID, year, month int) (budgets []*Budget, err error) {
	rows, err := d.dbpool.Query(
		context.Background(),
		"select * from budgets where tag_id=$1 and extract(year from date) = $2 and extract(month from date) = $3",
		tagId.String(),
		year,
		month,
	)
	if err != nil {
		return nil, fmt.Errorf("could not list budgets: %w", err)
	}

	for rows.Next() {
		var budget Budget
		err := rows.Scan(
			&budget.ID,
			&budget.Label,
			&budget.Amount,
			&budget.Date,
			&budget.TagID,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan budget: %w", err)
		}

		budgets = append(budgets, &budget)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate budgets: %w", err)
	}

	return
}

func (d *Datastore) ListTags() (tags []*Tag, err error) {
	rows, err := d.dbpool.Query(context.Background(), "select * from tags")
	if err != nil {
		return nil, fmt.Errorf("could not list tags: %w", err)
	}

	for rows.Next() {
		var tag Tag
		err := rows.Scan(
			&tag.ID,
			&tag.Label,
			&tag.Description,
			&tag.Icon,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan tags: %w", err)
		}

		tags = append(tags, &tag)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate tags: %w", err)
	}

	return
}

func (d *Datastore) ListExpenses(year, month int) (expenses []*Expense, err error) {
	rows, err := d.dbpool.Query(
		context.Background(),
		"select * from expenses where extract(year from date) = $1 and extract(month from date) = $2",
		year,
		month,
	)
	if err != nil {
		return nil, fmt.Errorf("could not list expenses: %w", err)
	}

	for rows.Next() {
		var expense Expense
		err := rows.Scan(
			&expense.ID,
			&expense.Label,
			&expense.Amount,
			&expense.Date,
			&expense.BudgetID,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan expense: %w", err)
		}

		expenses = append(expenses, &expense)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate expenses: %w", err)
	}

	return
}

func (d *Datastore) ListExpensesForBudget(year, month int, budget string) (expenses []*Expense, err error) {
	rows, err := d.dbpool.Query(
		context.Background(),
		"select * from expenses where budget=$1 and extract(year from date) = $2 and extract(month from date) = $3",
		budget,
		year,
		month,
	)
	if err != nil {
		return nil, fmt.Errorf("could not list expenses: %w", err)
	}

	for rows.Next() {
		var expense Expense
		err := rows.Scan(
			&expense.ID,
			&expense.Label,
			&expense.Amount,
			&expense.Date,
			&expense.BudgetID,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan expense: %w", err)
		}

		expenses = append(expenses, &expense)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate expenses: %w", err)
	}

	return
}

func (d *Datastore) ListIncomes(year, month int) (incomes []*Income, err error) {
	rows, err := d.dbpool.Query(
		context.Background(),
		"select * from incomes where extract(year from date) = $1 and extract(month from date) = $2", year, month,
	)
	if err != nil {
		return nil, fmt.Errorf("could not list incomes: %w", err)
	}

	for rows.Next() {
		var income Income
		err := rows.Scan(
			&income.ID,
			&income.Label,
			&income.Amount,
			&income.Date,
		)

		if err != nil {
			return nil, fmt.Errorf("could not scan income: %w", err)
		}

		incomes = append(incomes, &income)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("could not iterate incomes: %w", err)
	}

	return
}
